x-pack/plugins/actions/server/builtin_action_types/lib/send_email.ts

```diff
@@ -75,25 +81,60 @@ export async function sendEmail(logger: Logger, options: SendEmailOptions): Prom
 async function sendEmailWithExchange(
   logger: Logger,
   options: SendEmailOptions,
-  messageHTML: string
+  messageHTML: string,
+  connectorTokenClient: ConnectorTokenClient
 ): Promise<unknown> {
-  const { transport, configurationUtilities } = options;
+  const { transport, configurationUtilities, connectorId } = options;
   const { clientId, clientSecret, tenantId, oauthTokenUrl } = transport;
-  // request access token for microsoft exchange online server with Graph API scope
 
-  const tokenResult = await requestOAuthClientCredentialsToken(
-    oauthTokenUrl ?? `${EXCHANGE_ONLINE_SERVER_HOST}/${tenantId}/oauth2/v2.0/token`,
-    logger,
-    {
-      scope: GRAPH_API_OAUTH_SCOPE,
-      clientId,
-      clientSecret,
-    },
-    configurationUtilities
-  );
+  let accessToken: string;
+
+  const connectorToken = await connectorTokenClient.get({ connectorId });
+  if (connectorToken === null || Date.parse(connectorToken.expiresAt) <= Date.now()) {
+    // request new access token for microsoft exchange online server with Graph API scope
+    const tokenResult = await requestOAuthClientCredentialsToken(
+      oauthTokenUrl ?? `${EXCHANGE_ONLINE_SERVER_HOST}/${tenantId}/oauth2/v2.0/token`,
+      logger,
+      {
+        scope: GRAPH_API_OAUTH_SCOPE,
+        clientId,
+        clientSecret,
+      },
+      configurationUtilities
+    );
+    accessToken = `${tokenResult.tokenType} ${tokenResult.accessToken}`;
+
+    // try to update connector_token SO
+    try {
+      if (connectorToken === null) {
```

Comment:
So one thing I'm seeing (and not sure the right fix for it) is if I try and manually update the saved object to expire in the past (by a `PUT` in dev tools), I'm invalidating the AAD and see an error like this at runtime:
```
[2021-12-02T09:47:03.755-05:00][ERROR][plugins.encryptedSavedObjects] Failed to decrypt "token" attribute: Unsupported state or unable to authenticate data
[2021-12-02T09:47:03.755-05:00][ERROR][plugins.actions] Failed to decrypt connector_token for connectorId "f96303e0-5206-11ec-93e1-b1beb34a39a9" and tokenType: "access_token". Error: Unable to decrypt attribute "token"
```

That error will result in `connectorToken = null` which means another saved object is created, and the one with the decrypt error is basically useless but sticks around indefinitely.

Is it possible to explicitly detect this scenario and delete the saved object or something? WDYT?

Reply:
This is probably an uncommon issue, but definitely need to be solved. I think we still need to log what happened during the update try and in case of error catch just delete this `connector_token` record.
In that case we won't overcomplicate the logic and won't keep corrupted tokens for the majority of the cases.
Added this in the commit

As a result of testing I don't see multiple connector tokens any more.

Reply:
Sounds good, thanks!